flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/source/reader/TestIcebergSourceReader.java

```diff
@@ -0,0 +1,191 @@
+
+  private final GenericAppenderFactory appenderFactory;
+
+  public TestIcebergSourceReader() {
+    this.appenderFactory = new GenericAppenderFactory(TestFixtures.SCHEMA);
+  }
+
+  @Test
+  public void testReaderMetrics() throws Exception {
+    TestingReaderOutput<RowData> readerOutput = new TestingReaderOutput<>();
+    TestingMetricGroup metricGroup = new TestingMetricGroup();
+    TestingReaderContext readerContext = new TestingReaderContext(new Configuration(), metricGroup);
+    IcebergSourceReader reader = createReader(metricGroup, readerContext);
+    reader.start();
+
+    testOneSplitFetcher(reader, readerOutput, metricGroup, 1);
+    testOneSplitFetcher(reader, readerOutput, metricGroup, 2);
+  }
+
+  private void testOneSplitFetcher(
+      IcebergSourceReader reader,
+      TestingReaderOutput<RowData> readerOutput,
+      TestingMetricGroup metricGroup,
+      int expectedCount)
+      throws Exception {
+    long seed = expectedCount;
+    // Each split should contain only one file with one record
+    List<List<Record>> recordBatchList =
+        ReaderUtil.createRecordBatchList(seed, TestFixtures.SCHEMA, 1, 1);
+    CombinedScanTask task =
+        ReaderUtil.createCombinedScanTask(
+            recordBatchList, TEMPORARY_FOLDER, FileFormat.PARQUET, appenderFactory);
+    IcebergSourceSplit split = IcebergSourceSplit.fromCombinedScanTask(task);
+    reader.addSplits(Arrays.asList(split));
+
+    while (readerOutput.getEmittedRecords().size() < expectedCount) {
+      reader.pollNext(readerOutput);
+    }
+
+    Assert.assertEquals(expectedCount, readerOutput.getEmittedRecords().size());
+    TestHelpers.assertRowData(
+        TestFixtures.SCHEMA,
+        recordBatchList.get(0).get(0),
+        readerOutput.getEmittedRecords().get(expectedCount - 1));
+    Assert.assertEquals(expectedCount, metricGroup.counters.get("assignedSplits").getCount());
+
+    // One more poll will get null record batch.
+    // That will finish the split and cause split fetcher to be closed due to idleness.
+    // Then next split will create a new split reader.
+    reader.pollNext(readerOutput);
+    Assert.assertEquals(expectedCount, metricGroup.counters.get("finishedSplits").getCount());
+  }
+
+  private IcebergSourceReader createReader(
+      MetricGroup metricGroup, SourceReaderContext readerContext) {
+    IcebergSourceReaderMetrics readerMetrics =
+        new IcebergSourceReaderMetrics(metricGroup, "db.tbl");
+    RowDataReaderFunction readerFunction =
+        new RowDataReaderFunction(
+            new Configuration(),
+            TestFixtures.SCHEMA,
+            TestFixtures.SCHEMA,
+            null,
+            true,
+            new HadoopFileIO(new org.apache.hadoop.conf.Configuration()),
+            new PlaintextEncryptionManager());
+    return new IcebergSourceReader<>(readerMetrics, readerFunction, readerContext);
+  }
+
+  private static class TestingMetricGroup extends UnregisteredMetricsGroup
```

Comment:
Just a question: Is it possible that this class would be useful for other tests later? If so, maybe putting it to the `org.apache.flink.connector.testutils.source.reader` package would be better. OTOH if we think this will never be used anywhere else, then this is a good place.

Reply:
Don't know if it will be used by other tests or not. I typically like to refactor later when the needs come. But I see the `TestingMetricGroup` has been big enough to be a stand-alone class. Hence I moved it to its own file. it is package private for now.